src/main/java/io/openliberty/tools/common/plugins/util/BinaryScannerUtil.java

```diff
@@ -0,0 +1,263 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.openliberty.tools.common.plugins.util;
+
+import java.io.File;
+import java.lang.reflect.InvocationTargetException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public abstract class BinaryScannerUtil {
+
+    public static final String BINARY_SCANNER_CONFLICT_MESSAGE1 = "A working set of features could not be generated due to conflicts " +
+            "between configured features and the application's API usage: %s. Review and update your server configuration and " +
+            "application to ensure they are not using conflicting features and APIs from different levels of MicroProfile, " +
+            "Java EE, or Jakarta EE. Refer to the following set of suggested features for guidance: %s";
+    public static final String BINARY_SCANNER_CONFLICT_MESSAGE2 = "A working set of features could not be generated due to conflicts " +
+            "between configured features: %s. Review and update your server configuration to ensure it is not using conflicting " +
+            "features from different levels of MicroProfile, Java EE, or Jakarta EE. Refer to the following set of " +
+            "suggested features for guidance: %s";
+    public static final String BINARY_SCANNER_CONFLICT_MESSAGE3 = "A working set of features could not be generated due to conflicts " +
+            "in the application’s API usage: %s. Review and update your application to ensure it is not using conflicting APIs " +
+            "from different levels of MicroProfile, Java EE, or Jakarta EE.";
+    public static final String BINARY_SCANNER_CONFLICT_MESSAGE4 = "[None available]"; // format should match JVM Set.toString()
+
+    public abstract void debug(String message);
+    public abstract void debug(String message, Throwable e);
+    public abstract void error(String message);
+    public abstract void warn(String message);
+    public abstract void info(String message);
+
+    // The jar file containing the binary scanner code
+    private File binaryScanner;
+    private URLClassLoader binaryScannerClassLoader = null;
+    
+    public BinaryScannerUtil(File binaryScanner) {
+        this.binaryScanner = binaryScanner;
+    }
+
+    public Set<String> runBinaryScanner(Set<String> currentFeatureSet, List<String> classFiles, Set<String> allClassesDirectories,
```

Comment:
It would be good to guard against an infinite loop since this method is called recursively. There should be a condition that could be checked to bail out at the beginning of the method if the calls have gone more levels than expected. You can open a separate issue to address after the snapshot.

Reply:
Looking at this.

Reply:
Issue opened here: https://github.com/OpenLiberty/ci.maven/issues/1341

Reply:
Recorded as https://github.com/OpenLiberty/ci.maven/issues/1341 and fixed in https://github.com/OpenLiberty/ci.common/pull/328